src/pstore/pstore.c

```diff
@@ -0,0 +1,582 @@
+    if ((r == 0) && remove_file && arg_unlink)
+    {
+        unlink(ifdpath);
+    }
+
+    return r;
+}
+
+static void write_dmesg (const char *dmesg, ssize_t size, const char *id) {
+    if ((dmesg != NULL) && (size > 0))
+    {
+        _cleanup_free_ char *ofdpath;
+        int ofd;
+
+        log_info("Record ID %s\n", id);
+
+        ofdpath = path_join(arg_archivedir, id, "dmesg.txt");
+
+        ofd = open(ofdpath, O_CREAT|O_TRUNC|O_WRONLY, S_IRUSR|S_IWUSR|S_IRGRP|S_IROTH/*0644*/);
+        if (ofd < 0)
+        {
+            log_error_errno(errno, "open(%s): %m\n", ofdpath);
+        }
+        else
+        {
+            ssize_t wr;
+            wr = write(ofd, dmesg, size);
+            if (wr != size)
+            {
+                log_error_errno(errno, "write(%s): %m\n", ofdpath);
+            }
+        }
+        close(ofd);
+    }
+}
+
+/*
+ * PStore content rule handling
+ */
+typedef struct rules_s
+{
+    void (*handler)(struct rules_s *rule, dirent_ct *dc);
+    const char *handlerName;
+    const char *pattern;
+} rules_t;
+static void dmesg_handler (struct rules_s *rule, dirent_ct *dc);
+static void move_handler (struct rules_s *rule, dirent_ct *dc);
+rules_t rules[] =
+{
+    /* from Linux/fs/pstore/inode.c
+     * the %s is record->psi->name (eg efi, erst)
+     * the %llu is record->id
+     */
+    { dmesg_handler, "dmesg", "dmesg-*", },     /* PSTORE_TYPE_DMESG: "dmesg-%s-%llu%s"     */
+    { move_handler, "console", "console-*", },  /* PSTORE_TYPE_CONSOLE: "console-%s-%llu"   */
+    { move_handler, "pmsg", "pmsg-*", },        /* PSTORE_TYPE_PMSG: "pmsg-%s-%llu"         */
+    { move_handler, "ftrace", "ftrace-*", },    /* PSTORE_TYPE_FTRACE: "ftrace-%s-%llu"     */
+    { move_handler, "mce", "mce-*", },          /* PSTORE_TYPE_MCE: "mce-%s-%llu"           */
+    { move_handler, "ppcrtas", "rtas-*", },     /* PSTORE_TYPE_PPC_RTAS: "rtas-%s-%llu"     */
+    { move_handler, "ppcof", "powerpc-ofw-*", },/* PSTORE_TYPE_PPC_OF: "powerpc-ofw-%s-%llu"*/
+    { move_handler, "ppccommon", "powerpc-common-*", }, /* PSTORE_TYPE_PPC_COMMON: "powerpc-common-%s-%llu" */
+    { move_handler, "ppcopal", "powerpc-opal-*", },     /* PSTORE_TYPE_PPC_OPAL: "powerpc-opal-%s-%llu"     */
+    { move_handler, "unknown", "unknown-*", },      /* PSTORE_TYPE_UNKNOWN: "unknown-%s-%llu"   */
+};
+
+static void dmesg_handler (struct rules_s *rule, dirent_ct *dc) {
+    /* dmesg-* files recorded on this list */
+    static dirent_ct *dmesgfiles = NULL;
+
+    if (dc != NULL) {
+        /* Collect the file for later processing */
+        dc->is_binary = false;
+        LIST_APPEND(dlcl, dmesgfiles, dc);
+    }
+    else {
+        /* End of files, move files, reconstruct dmesg.txt */
+        char *dmesg = NULL, *dmesg2;
+        ssize_t dmesgsize = 0;
+        char id1[ENTRYSZ], id2[ENTRYSZ], *id;
+        char *currentid;
+        char *p;
+
+        if (dmesgfiles == NULL) return;
```

Comment:
we usually don't bother with `== NULL` and rely on C's downgrade-to-bool feature for pointers. (but only for pointers besides everything that is a boolean anyway, see coding style)

Reply:
I don't see the '== NULL' addressed in coding style doc, but I've changed it accordingly.

Reply:
indeed. this was missing. will add.

Reply:
Posted #12875 about this.

Reply:
Fwiw, I peeked at the change to coding style on this topic, and I know that I personally would benefit from your "do this" and "dont do this" examples with respect to the use of NULL pointer checking... Or said differently, I probably would not have made the connection of NULL pointer checking and boolean given the description. $0.02.